packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmPropertyType.kt

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.schema
+
+import kotlin.reflect.KClass
+
+/**
+ * A [RealmPropertyType] describes the type of a specific property in the object model.
+ */
+sealed interface RealmPropertyType {
+    /**
+     * The type that is used when storing the property values in the realm.
+     */
+    val storageType: RealmStorageType
+
+    /**
+     * Indicates whether the storage element can be `null`.
+     */
+    val isNullable: Boolean
+
+    companion object {
+        // TODO Not as good as RealmPropertyType::class.sealedClasses as this has to be manually
+        //  adjusted, but since KClass<T>.sealedClasses is only available for JVM this is the next
+        //  best thing (at least uncovered until now ... without writing a compiler plugin) that
+        //  allows to define the options centrally and use it to verify exhaustiveness in tests.
+        //  JUST DON'T FORGET TO UPDATE ON WHEN ADDING NEW SUBCLASSES :see_no_evil:
+        //  We could do a JVM test that verifies that it is exhaustive :thinking:
+        internal val subTypes: Set<KClass<out RealmPropertyType>> = setOf(ValuePropertyType::class, ListPropertyType::class)
+    }
+}
+
+/**
+ * A [ValuePropertyType] describes singular value properties.
+ */
+data class ValuePropertyType(
```

Comment:
I like this name, but after talking to @edualonso , we might have small design headache around queries on objects and primitive lists that will require us to have two implementations. In that case, we should probably align whatever we end up calling the query classes with this.

Reply:
Sigh... 😢